### PR TITLE
[SPARK-5929] Pyspark: Register a pip requirements file with spark_context

### PR DESCRIPTION
**Pull request by buckhx**
Ships all packages in the requirements file by installing them locally via pip and then ships the packages to the workers via addPyFile

### CODE REVIEW

File: python/pyspark/context.py [hunk 1]
```diff
@@ -65,8 +65,9 @@ class SparkContext(object):
     _python_includes = None  # zip and egg files that need to be added to PYTHONPATH
 
     def __init__(self, master=None, appName=None, sparkHome=None, pyFiles=None,
-                 environment=None, batchSize=0, serializer=PickleSerializer(), conf=None,
-                 gateway=None, jsc=None, profiler_cls=BasicProfiler):
+                 requirementsFile=None, environment=None, batchSize=0,
```
Comment (davies) [inline python/pyspark/context.py]: Insert a parameter in the middle will break compatibility, we should put it in the end.

File: python/pyspark/context.py [hunk 2]
```diff
@@ -710,6 +717,33 @@ def addPyFile(self, path):
...
+        import pip
+        import tarfile
+        import tempfile
+        import uuid
+        tar_dir = tempfile.mkdtemp()
+        try:
+            for req in pip.req.parse_requirements(path, session=uuid.uuid1()):
+                if not req.check_if_exists():
+                    pip.main(['install', req.req.__str__()])
+                mod = importlib.import_module(req.name) # Can throw ImportError
```
Comment (davies) [inline python/pyspark/context.py]: import lib is not available in Python 2.6

**Comment by davies** [comment]
Good idea, this is a useful feature to have. Could you also a argument for spark-submit, (for example, --pip or --py-requirements). Also, could you add a test for it?

**Comment by davies** [comment]
Jenkins, Ok to test.

**Comment by buckheroux** [comment]
Great, will get to that this week and add tests.

Do you think bundling importlib with pyspark is reasonable? Or is finding another way to track down the local package the way to go?

**Comment by davies** [comment]
I think we can go with `__import__`, it's similar to importlib

**Comment by buckheroux** [comment]
Changed importlib to __import__ and moved most of the requirements up. I left the pip requirement in the method itself as to not require the system has pip unless that method is used. Also added a test that tests the import of a library not found on most systems https://github.com/buckheroux/QuadKey.

**Comment by buckheroux** [comment]
Any traction on this?

**Comment by robcowie** [comment]
This could be very useful. One question; Does this behave with namespace packages?

**Comment by buckheroux** [comment]
@robcowie I wouldn't think so. Do you have a good example of a pypi package that I could try to install that has a namespace? I can try and support it if possible.

**Comment by andrewor14** [comment]
Jenkins slow test please

**Comment by buckheroux** [comment]
Just saw the test failure, looking into it now

File: python/pyspark/context.py [hunk 3]
```diff
@@ -67,8 +70,9 @@ class SparkContext(object):
     PACKAGE_EXTENSIONS = ('.zip', '.egg', '.jar')
 
     def __init__(self, master=None, appName=None, sparkHome=None, pyFiles=None,
-                 environment=None, batchSize=0, serializer=PickleSerializer(), conf=None,
-                 gateway=None, jsc=None, profiler_cls=BasicProfiler):
+                 environment=None, batchSize=0, serializer=PickleSerializer(),
+                 conf=None, gateway=None, jsc=None, profiler_cls=BasicProfiler,
+                 requirementsFile=None):
```
Comment (davies) [inline python/pyspark/context.py]: We already have two much parameters here, Can we just have `addRequirementsFile`?

File: python/pyspark/context.py [hunk 4]
```diff
@@ -711,6 +721,30 @@ def addPyFile(self, path):
...
+        """
+        import pip
+        tar_dir = tempfile.mkdtemp()
+        try:
+            for req in pip.req.parse_requirements(path, session=uuid.uuid1()):
+                if not req.check_if_exists():
+                    pip.main(['install', req.req.__str__()])
+                mod = __import__(req.name)
+                mod_path = mod.__path__[0]
+                tar_path = os.path.join(tar_dir, req.name+'.tar.gz')
```
Comment (davies) [inline python/pyspark/context.py]: Is possible that we can re-use the packaged file?

**Comment by andrewor14** [comment]
@buckheroux I believe the test failure is not related. @JoshRosen is looking into a hot fix.

**Comment by davies** [comment]
@buckheroux Do you think that we should also support this in `bin/spark-submit`? For example, we should accept `.txt` file (via --py-files) as dependencies.

File: python/pyspark/context.py [hunk 5]
```diff
@@ -711,6 +721,30 @@ def addPyFile(self, path):
...
+        Raises ImportError if the requirement can't be found
+        """
+        import pip
+        tar_dir = tempfile.mkdtemp()
+        try:
+            for req in pip.req.parse_requirements(path, session=uuid.uuid1()):
+                if not req.check_if_exists():
+                    pip.main(['install', req.req.__str__()])
+                mod = __import__(req.name)
+                mod_path = mod.__path__[0]
```
Comment (buckheroux) [inline python/pyspark/context.py]: Worried about this working across different types of modules (namespace packages etc)

File: python/pyspark/context.py (hunk 4 above)
Comment (buckheroux) [inline python/pyspark/context.py]: Hmm good thought. I'll look into this.

**Comment by buckheroux** [comment]
@andrewor14 Sounds good.

@davies It's probably a good idea to support a CLI arg for the requirements file. I'll work on this as well.

File: python/pyspark/context.py (hunk 3 above)
Comment (buckheroux) [inline python/pyspark/context.py]: I was trying to match the pyFiles interface as closely as possible, although I do see where you are coming from. I'll re-evaluate having the requirementsFile kwarg after add a requirements file arg to the CLI.

**Comment by JoshRosen** [comment]
@andrewor14 the test failure in the `dev/run-tests` doctest are because we did a shallow clone of the Spark repository into the fresh Jeknins workspace. I think that this is an ephemeral problem that will resolve itself after retesting. We could also disable / remove that test for now or skip it in Jenkins.

**Comment by JoshRosen** [comment]
I've disabled shallow cloning in SlowSparkPullRequestBuilder, so let's try this again...

**Comment by JoshRosen** [comment]
jenkins slow test please

File: python/pyspark/context.py (hunk 5 above)
Comment (buckhx) [inline python/pyspark/context.py]: Indeed mod.__path__[0] does not work for namespace packages. Working on supporting them by bundling all paths in mod.__path__ into a single tar.

**Comment by buckhx** [comment]
The tests fail because one of them attempts to pip install a package and doesn't have permissions to do so. Is there a way to enable that? Or just leave the pip installing piece untested?

**Comment by buckhx** [comment]
Looking to add support for namespaces, which I have tested locally and confirmed to work. I was also thinking of exposing an add_package function that would add local packages to the spark context.

**Comment by buckhx** [comment]
@davies removed the requirements file from the context constructor. I looked into add --py-requirements to spark-submit, but it looked a bit more in depth than I was thinking it would be. I think it makes sense to add that as a separate commit/PR. 

Also, when testing I import a package from pip, but the Spark Jenkins test errors because it doesn't allow for access to the global pypi server. Any thoughts?

**Comment by davies** [comment]
cc @JoshRosen @sknapp who are more familiar with Jenkins

**Comment by DoHe** [comment]
For the test you could also add a local egg file to the repository and install this instead of relying on the external pypi server.

**Comment by rxin** [comment]
I'm going to close this pull request. If this is still relevant and you are interested in pushing it forward, please open a new pull request. Thanks!

**Comment by buckhx** [comment]
This is a pretty big pain when using pyspark (adding modules to workers) and should definitely be included. I'll open a new pull request and to push this through.

**Comment by jurriaan** [comment]
Yeah, this is quite a useful feature :+1:

**Pull request closed without merging**